### PR TITLE
Merge benchmark

### PR DESCRIPTION
**Pull request by madsbk**
Added a cuDF merge benchmark based on @rjzamora's notebook: https://gist.github.com/rjzamora/0ffc35c19b5180ab04bbf7c793c45955

```$ local_merge.py --help
usage: local_merge.py [-h] [-d DEVS] [-p {tcp,ucx}] [-c n]
                      [--ignore-size nbytes] [--frac-match FRAC]
                      [--no-pool-allocator]

Merge (dask/cudf) on LocalCUDACluster benchmark

optional arguments:
  -h, --help            show this help message and exit
  -d DEVS, --devs DEVS  GPU devices to use (default "0").
  -p {tcp,ucx}, --protocol {tcp,ucx}
                        The communication protocol to use.
  -c n, --chunk-size n  Chunk size per worker (default 1_000_000)
  --ignore-size nbytes  Ignore messages smaller than this (default "1 MB")
  --frac-match FRAC     Fraction of rows that matches (default 0.3)
  --no-pool-allocator   Disable the RMM memory pool
```

#### UCX
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | ucx
Device(s)   | 1,2
==========================
Total time  | 8.57 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(01,02)     | 2.48 GB/s 3.21 GB/s 3.32 GB/s (5.20 GB)
(02,01)     | 2.45 GB/s 3.12 GB/s 3.26 GB/s (5.20 GB)
```

#### TCP
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | tcp
Device(s)   | 1,2
==========================
Total time  | 30.10 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(01,02)     | 136.55 MB/s 143.48 MB/s 150.41 MB/s (2.80 GB)
(02,01)     | 142.75 MB/s 154.91 MB/s 167.08 MB/s (2.80 GB)
```
Going from TCP to UCX we see a **3.5 times** speedup (wall-clock) and more than **20 times** higher bandwidth. 
Notice, that the total amount of data send in TCP is only 2.8 GB (one-way) whereas in UCX it is  5.20 GB. This is not always the case, some very few times UCX also only communicated 2.8 GB. 
I don't know what happens but my guess is that the data distribution by `distributed` is sensitive to network timings. @mrocklin @quasiben? 

#### UCX
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | ucx
Device(s)   | 4,5,6,7
==========================
Total time  | 18.97 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(04,05)     | 725.28 MB/s 881.78 MB/s 1.04 GB/s (1.40 GB)
(04,06)     | 718.94 MB/s 877.93 MB/s 1.31 GB/s (3.80 GB)
(04,07)     | 968.64 MB/s 1.26 GB/s 1.55 GB/s (1.40 GB)
(05,04)     | 676.29 MB/s 749.17 MB/s 1.66 GB/s (3.80 GB)
(05,06)     | 709.26 MB/s 892.33 MB/s 1.08 GB/s (1.40 GB)
(05,07)     | 935.18 MB/s 1.30 GB/s 1.66 GB/s (1.40 GB)
(06,04)     | 613.71 MB/s 662.51 MB/s 711.31 MB/s (1.40 GB)
(06,05)     | 679.15 MB/s 827.13 MB/s 975.12 MB/s (1.40 GB)
(06,07)     | 1.15 GB/s 1.70 GB/s 2.19 GB/s (3.80 GB)
(07,04)     | 1.03 GB/s 1.40 GB/s 1.78 GB/s (1.40 GB)
(07,05)     | 896.32 MB/s 1.24 GB/s 1.81 GB/s (3.80 GB)
(07,06)     | 638.65 MB/s 723.34 MB/s 808.03 MB/s (1.40 GB)
```
#### TCP
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | tcp
Device(s)   | 4,5,6,7
==========================
Total time  | 63.81 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(04,05)     | 63.66 MB/s 96.47 MB/s 143.60 MB/s (2.80 GB)
(04,06)     | 31.92 MB/s 36.21 MB/s 40.50 MB/s (1.40 GB)
(04,07)     | 42.46 MB/s 49.54 MB/s 79.11 MB/s (2.10 GB)
(05,04)     | 74.86 MB/s 88.78 MB/s 104.15 MB/s (4.50 GB)
(05,06)     | 87.26 MB/s 100.29 MB/s 117.02 MB/s (2.10 GB)
(05,07)     | 32.12 MB/s 32.12 MB/s 32.12 MB/s (699.95 MB)
(06,04)     | 36.48 MB/s 36.48 MB/s 36.48 MB/s (700.11 MB)
(06,05)     | 75.25 MB/s 89.02 MB/s 102.79 MB/s (1.40 GB)
(06,07)     | 89.11 MB/s 94.32 MB/s 108.84 MB/s (3.80 GB)
(07,04)     | 46.27 MB/s 55.91 MB/s 68.64 MB/s (2.10 GB)
(07,05)     | 36.90 MB/s 36.95 MB/s 37.00 MB/s (1.40 GB)
(07,06)     | 113.82 MB/s 155.57 MB/s 182.52 MB/s (2.80 GB)
```
Again, with four GPUs we see similar results: **3.4 speedup** (wall-clock) and more than **10 times** higher bandwidth in most cases.
Notice, the wall-clock does not goes down compared to the two-gpu case because the chunk-size is *per* worker.

### CODE REVIEW

File: dask_cuda/benchmarks/local_merge.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    ).persist()
+    ddf_other = get_random_ddf(
+        args.chunk_size, n_workers, args.frac_match, "other"
+    ).persist()
+
+    # Lazy merge/join operation
+    ddf_join = ddf_base.merge(ddf_other, on=["key"], how="inner")
+
+    t1 = clock()
+    wait(ddf_join.compute())
```
Comment (mrocklin) [inline dask_cuda/benchmarks/local_merge.py]:
```suggestion
    wait(ddf_join.persist())
```

If you call compute then you bring everything back to the local process, and a single GPU.  If you call persist then the data stays distributed across all of the workers.

File: dask_cuda/benchmarks/local_merge.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    n_workers = len(args.devs.split(","))
+
+    # Set up workers on the local machine
+    cluster = LocalCUDACluster(
+        protocol=args.protocol, n_workers=n_workers, CUDA_VISIBLE_DEVICES=args.devs,
+    )
+    client = Client(cluster)
+
+    if args.no_pool_allocator:
+        client.run(lambda: cudf.set_allocator("default", pool=False))
```
Comment (mrocklin) [inline dask_cuda/benchmarks/local_merge.py]: Small style suggestion

```suggestion
        client.run(cudf.set_allocator, "default", pool=False)
```

File: dask_cuda/benchmarks/local_merge.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    else:
+        client.run(lambda: cudf.set_allocator("default", pool=True))
```
Comment (mrocklin) [inline dask_cuda/benchmarks/local_merge.py]:
```suggestion
        client.run(cudf.set_allocator, "default", pool=True)
```

File: dask_cuda/benchmarks/local_merge.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        ]
+        for (w1, w2), v in bandwidths.items()
+    }
+    total_nbytes = {
+        (
+            cluster.scheduler.workers[w1].name,
+            cluster.scheduler.workers[w2].name,
+        ): format_bytes(sum(nb))
+        for (w1, w2), nb in total_nbytes.items()
+    }
```
Comment (mrocklin) [inline dask_cuda/benchmarks/local_merge.py]: You might also want to look at `cluster.scheduler.bandwidth_workers`

**Comment by mrocklin** [comment]
It's really nice to see benchmarks on the full cudf join problem.  

> This is not always the case, some very few times UCX also only communicated 2.8 GB.
I don't know what happens but my guess is that the data distribution by distributed is sensitive to network timings. @mrocklin @quasiben?

That doesn't sound right to me.  One issue might be the compute call I pointed to above.  Past that I think that we'll need better visibility into communications.  I'll see if I can expose some more state there.

**Comment by madsbk** [comment]
> That doesn't sound right to me. One issue might be the compute call I pointed to above. Past that I think that we'll need better visibility into communications. I'll see if I can expose some more state there.

Even after `.compute() => .persist()`, I still see a 2.8 GB approx. every 5 runs.

**Comment by madsbk** [comment]
BTW, @mrocklin when I try to use the profile in dashboard I get a `TypeError: 'coroutine' object is not subscriptable` error. Is that something you have seen before?

<details>

```
arks/local_merge.py -p ucx -d 1,2,3,4 -c 100000000
distributed.utils - ERROR - 'coroutine' object is not subscriptable
Traceback (most recent call last):
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/utils.py", line 663, in log_errors
    yield
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/dashboard/components/shared.py", line 380, in update
    data = profile.plot_data(self.state, profile_interval)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/profile.py", line 232, in plot_data
    traverse(state, 0, 1, 0)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/profile.py", line 195, in traverse
    if not state["count"]:
TypeError: 'coroutine' object is not subscriptable
tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOLoop object at 0x7fc3405f5358>>, <Future finished exception=TypeError("'coroutine' object is not subscriptable")>)
Traceback (most recent call last):
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
    ret = callback()
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/tornado/ioloop.py", line 767, in _discard_future_result
    future.result()
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/tornado/gen.py", line 748, in run
    yielded = self.gen.send(value)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/bokeh/server/session.py", line 70, in _needs_document_lock_wrapper
    result = yield yield_for_all_futures(func(self, *args, **kwargs))
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/bokeh/server/session.py", line 191, in with_document_locked
    return func(*args, **kwargs)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/bokeh/document/document.py", line 1127, in wrapper
    return doc._with_self_as_curdoc(invoke)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/bokeh/document/document.py", line 1113, in _with_self_as_curdoc
    return f()
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/bokeh/document/document.py", line 1126, in invoke
    return f(*args, **kwargs)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/bokeh/document/document.py", line 916, in remove_then_invoke
    return callback(*args, **kwargs)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/dashboard/components/shared.py", line 409, in <lambda>
    self.doc().add_next_tick_callback(lambda: self.update(prof, metadata))
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/bokeh/core/property/validation.py", line 97, in func
    return input_function(*args, **kwargs)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/dashboard/components/shared.py", line 380, in update
    data = profile.plot_data(self.state, profile_interval)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/profile.py", line 232, in plot_data
    traverse(state, 0, 1, 0)
  File "/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/distributed/profile.py", line 195, in traverse
    if not state["count"]:
TypeError: 'coroutine' object is not subscriptable
/home/nfs/mkristensen/anaconda3/envs/ucx-benchmark/lib/python3.7/site-packages/tornado/locks.py:536: RuntimeWarning: coroutine 'Scheduler.get_profile_metadata' was never awaited
  return self._block.acquire(timeout)
```
</details>

**Comment by pentschev** [comment]
I'm going to be the annoying grammar guy: I see you're using the term "nodes" when you refer to GPUs, I think generally that term is used when talking about multiple machines (e.g., dgx11 is a node and dgx12 is a node). Probably referring to GPUs as GPUs (or as devices) instead of nodes would leave less room for ambiguity. Sorry for being the annoying guy. :)

Overall, this PR looks great to me. I'm gonna start running them on a regular basis on our DGXs (probably nightly).

**Comment by mrocklin** [comment]
Hrm, I have not seen that before, but I was just mucking about in that code and it's not very well tested (UI code is hard).  I've reproduced and taking a look now.

**Comment by mrocklin** [comment]
The profile issue should be resolved in master now.

**Comment by madsbk** [comment]
**Wow**, the cuDF PR https://github.com/rapidsai/cudf/pull/3478, which fixes the [unnecessarily index deep copy issue](https://github.com/rapidsai/cudf/issues/3477) makes a huge different!
Notice, other changes might also have made an impact -- the following result is with fully updated repos on `dgx18`.

### Two GPUs on a DGX1

#### Tl;dr: 
Speedup going from TCP to UCX:
Wall-clock: **21.5 times!**
Bandwidth: **200 times!**

#### UCX
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | ucx
Device(s)   | 1,2
==========================
Total time  | 2.46 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(01,02)     | 13.33 GB/s 22.76 GB/s 24.38 GB/s (5.20 GB)
(02,01)     | 11.89 GB/s 19.80 GB/s 25.95 GB/s (5.20 GB)
```
Notice, peak is 50GB/s (two NVLINKs)

#### TCP
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | tcp
Device(s)   | 1,2
==========================
Total time  | 52.97 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(01,02)     | 100.58 MB/s 108.06 MB/s 113.09 MB/s (5.20 GB)
(02,01)     | 113.35 MB/s 122.18 MB/s 134.94 MB/s (5.20 GB)
```
### Four GPUs on a DGX1

#### UCX
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | ucx
Device(s)   | 0,1,2,3
==========================
Total time  | 2.92 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(00,01)     | 2.07 GB/s 2.64 GB/s 3.22 GB/s (1.40 GB)
(00,02)     | 2.18 GB/s 2.93 GB/s 3.67 GB/s (1.40 GB)
(00,03)     | 2.84 GB/s 4.14 GB/s 17.25 GB/s (3.80 GB)
(01,00)     | 2.80 GB/s 4.12 GB/s 8.02 GB/s (3.80 GB)
(01,02)     | 2.69 GB/s 3.81 GB/s 4.94 GB/s (1.40 GB)
(01,03)     | 2.20 GB/s 2.93 GB/s 3.65 GB/s (1.40 GB)
(02,00)     | 2.14 GB/s 2.60 GB/s 3.06 GB/s (1.40 GB)
(02,01)     | 2.47 GB/s 3.34 GB/s 4.22 GB/s (1.40 GB)
(02,03)     | 2.37 GB/s 3.05 GB/s 3.72 GB/s (1.40 GB)
(03,00)     | 2.39 GB/s 3.19 GB/s 4.00 GB/s (1.40 GB)
(03,01)     | 2.85 GB/s 4.14 GB/s 11.11 GB/s (3.80 GB)
(03,02)     | 2.47 GB/s 3.37 GB/s 4.27 GB/s (1.40 GB)
```

#### TCP
```
Merge benchmark
--------------------------
Chunk-size  | 100000000
Frac-match  | 0.3
Ignore-size | 1.05 MB
Protocol    | tcp
Device(s)   | 0,1,2,3
==========================
Total time  | 69.61 s
==========================
(w1,w2)     | 25% 50% 75% (total nbytes)
--------------------------
(00,01)     | 56.18 MB/s 63.24 MB/s 70.31 MB/s (1.40 GB)
(00,02)     | 50.41 MB/s 56.00 MB/s 69.10 MB/s (3.80 GB)
(00,03)     | 42.66 MB/s 44.67 MB/s 46.69 MB/s (1.40 GB)
(01,00)     | 40.70 MB/s 40.70 MB/s 40.70 MB/s (700.04 MB)
(01,02)     | 56.65 MB/s 60.60 MB/s 64.54 MB/s (1.40 GB)
(01,03)     | 79.28 MB/s 111.50 MB/s 114.57 MB/s (3.80 GB)
(02,00)     | 59.08 MB/s 82.82 MB/s 108.04 MB/s (4.50 GB)
(02,01)     | 67.86 MB/s 72.93 MB/s 101.29 MB/s (2.80 GB)
(02,03)     | 62.73 MB/s 79.31 MB/s 95.89 MB/s (1.40 GB)
(03,00)     | 47.00 MB/s 53.38 MB/s 59.76 MB/s (1.40 GB)
(03,01)     | 45.98 MB/s 46.76 MB/s 47.54 MB/s (1.40 GB)
(03,02)     | 43.86 MB/s 51.91 MB/s 81.22 MB/s (3.80 GB)
```

**Comment by madsbk** [comment]
@pentschev, I think this is ready to be merged.

**Comment by pentschev** [comment]
Those numbers looks awesome, especially when compared to yesterday's!

Thanks @madsbk for this benchmark!

**Comment by rjzamora** [comment]
Awesome! Thanks @madsbk!

**Comment by mrocklin** [comment]
Nice results :)

I'm curious what the current bottlenecks are. Are we compute bound? communication bound?  If so, where?

If you're on distributed master and can add the following lines that would be welcome:

```python
client.profile(filename="comptuation.html")
client.profile(server=True, "communication.html")
```

And add the following to your config:

```yaml
distributed:
  comm:
    offload: False
```

That will give you a couple of HTML files that you'll have to use a web server to inspect, either `python -m http.server` or put them in some `gh-pages` branch on github, or just copy them here and I'll put them up on some web server.

**Comment by datametrician** [comment]
I'd also like to see this for 8x GPU to see what happens when V100 aren't fully connected.

**Comment by mrocklin** [comment]
An alternative approach for sharing benchmarking results: https://github.com/dask/distributed/pull/3282

**Comment by madsbk** [comment]
@mrocklin, here are the profile files: [profile.zip](https://github.com/rapidsai/dask-cuda/files/3903002/profile.zip)

**Comment by mrocklin** [comment]
Thanks @madsbk !

The communication.html file tracks the workers' administrative threads.  It shows that the worker wasn't doing much other than initializing the RMM pool at the beginning.  We're not seeing any time spent serializing or communicating here.  Did you set `distributed.comm.offload: False`?  That should make sure that all of the serialization/deserialization happens on the main thread as well so that it gets included.

For computation it appears that we're spending a lot of time in a cudf index.copy call, which is being called from the `drop` method.  This seems like the sort of thing that shouldn't require a full copy, but I'm not sure.  @kkraus14 you might want to take a look.  There is also a surprising amount of time spent in `Series.__mod__`.

After looking at these results I'm now also curious about the task stream plot to see if we're communication or computation bound.  I've also merged in the `performance_report` PR.  That may make it easier to bundle together these results together in the future.

**Comment by jakirkham** [comment]
> For computation it appears that we're spending a lot of time in a cudf index.copy call, which is being called from the drop method. This seems like the sort of thing that shouldn't require a full copy, but I'm not sure. @kkraus14 you might want to take a look.

Was just raising an issue about that. ( https://github.com/rapidsai/cudf/issues/3493 ) 😆
